src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs

```diff
@@ -115,11 +115,15 @@ private bool CheckGlobalConditions(Log log, DateTime lastCheckedAtUtc, UpToDateC
                 return log.Fail("FirstRun", "The up-to-date check has not yet run for this project. Not up-to-date.");
             }
 
-            string copyAlwaysItemPath = state.ItemsByItemType.SelectMany(kvp => kvp.Value).FirstOrDefault(item => item.CopyType == CopyType.CopyAlways).Path;
-
-            if (copyAlwaysItemPath != null)
+            foreach ((_, ImmutableArray<(string Path, string? Link, CopyType CopyType)> items) in state.ItemsByItemType)
```

Comment:
Records are heap allocated (struct records are proposed for future), so converting this would be a pessimization.

Records also generate a lot of backing code, increasing deployment size and JIT time.

A value tuple is a real type. You can make extension methods on them.

These aren't part of the public API.

Reply:
Interesting. Never heard of `pessimization` before. I was aware that records are just shorthand-declared classes with extra generated code. I generally think in 'ease of use/understanding' over 'memory allocation'. I wonder if using structs would create unintended side effects, so I'm not going to go down that rabbit hole. I didn't know you can make extension methods for value tuples. I'm aware these are not public API surface, but generally, it is a code smell to expose tuple-based interfaces (not C# interface, but the API-meaning of interface) outside of a class itself. It creates difficulty in following logic when you have a type always defined by shape with no name associated with it. Humans like to name things. 😝 TANGENT WARNING: It is the same reason (I believe) you prefer class name usage over `var` for variable declaration. I personally prefer `var` because it separates out pure logic structure from declaration as a defined code shape. My brain converts code to shapes (in a way) and `var` makes shape brain smile. 😂

Reply:
> It is the same reason (I believe) you prefer class name usage over var for variable declaration

In personal projects I use var everywhere. I'm just playing by house rules. Though honestly in VS I prefer explicit types as there are many, many more types and concepts in VS than in my personal projects, and it's harder to infer types from other symbols.